### ISSUE TITLE
Murmurhash algorithm results vary based on host endianness 

### ISSUE THREAD

**Issue by andrewsi-z** [explosion/murmurhash#26]
Hello,

I am seeing that murmurhash produces different results on platforms of different byte order. 
This will impact the results in cases where the model was trained on x86 and loaded / used on a big endian platform like s390x. 

So far, I've encountered this twice:
   - murmurhash3 in MurmurHash3_x64_128. 
   - murmurhash2 in MurmurHash64A.

In the context of spacy/thinc usage, this produces a result that can be somewhat unpredictable. With the fixes in https://github.com/explosion/thinc/pull/559 , I was still observing a difference in the resulting analysis of certain sentences, while others matched exactly.  

I have added fixes to the issues encountered so far in: https://github.com/andrewsi-z/murmurhash
- murmurhash3 fixes based on existing node implementation (this is used in various packages):
   - https://github.com/hideo55/node-murmurhash3/
- murmurhash2 hash64a based on mumurhash3 design.

On s390x or other big endian platforms, simply cloning and pip installing https://github.com/andrewsi-z/murmurhash is enough to replace/bypass the install of https://github.com/exposion/murmurhash. 

In terms of this issue and our discussions in the below related issues, are you interested in a PR for these and future endian relaed changes for the murmurhash implementation, or would you prefer to keep this as a separate repository?

Thank you!

See also:

https://github.com/explosion/thinc/pull/559
https://github.com/explosion/spaCy/issues/9428

**Comment by honnibal** [comment]
@andrewsi-z If the patch just adds some functions and tests, it looks like a good one to have in the main library, so sure open the PR, thanks!

**Comment by andrewsi-z** [comment]
@adrianeboyd I see you just merged a fix for murmurhash3. murmurhash2 still has issues with hash64A... Possibly other functions as well, but I do not see the use of those others. 

hash64A is used via hash64 cython and I see it exposed in thinc ngrams layer and in string.pyx (which is where I ran into issues with it).

I will pull murmurhash3 out of my fork. For murmurhash2, I followed the fix "model" used in the known murmurhash3 endian implementation here: https://github.com/hideo55/node-murmurhash3/

If you prefer, I can modify the murmurhash fix to more closely follow your model instead of what I used... It can be found here: https://github.com/andrewsi-z/murmurhash/blob/master/murmurhash/MurmurHash2.cpp

**Comment by adrianeboyd** [comment]
We thought that the implementation in the other library was a bit hard to follow. I think something similar for murmurhash2 would be okay, but I'm sure Matt will want to take a careful look.

**Comment by adrianeboyd** [comment]
Okay, I think this is coming along. For the hash-related parts (not the `asarray` parts), I think that these three PRs should make everything up through `thinc` endian-neutral:

* https://github.com/explosion/murmurhash/pull/27
* https://github.com/explosion/murmurhash/pull/28
* https://github.com/explosion/thinc/pull/565

Could you try it out and see if you run into any problems with `en_core_web_sm` + `BigEndianOps.asarray` (without `hash`)?

For `spacy`, I think there are (at least) two remaining concerns related to vectors:

* the vector data loading, where I think you would just need to add `ops.asarray` for the byte-swapping:
  * https://github.com/explosion/spaCy/blob/52b8c2d2e0241e1c515131c5e5f576d5dad65059/spacy/vectors.pyx#L593
  * https://github.com/explosion/spaCy/blob/52b8c2d2e0241e1c515131c5e5f576d5dad65059/spacy/vectors.pyx#L647
* murmurhash for floret vectors:
  * https://github.com/explosion/spaCy/pull/9735
  * https://github.com/explosion/floret/pull/14

**Comment by andrewsi-z** [comment]
Great! I applied the changes and did not see any issues - using `en_core_web_sm` + `BigEndianOps.asarray` resulted in output that matches an LE platform. 

Based on this, I removed hash from [https://github.com/andrewsi-z/thinc-bigendian-ops](https://github.com/andrewsi-z/thinc-bigendian-ops), with asarray remaining.

**Comment by adrianeboyd** [comment]
Glad to hear it! I think then the next question is how to handle the saved format, which I think we should probably always have as little endian. The problem is that `Ops.to_numpy` is also used within spacy in a few places for other reasons (numpy/cupy conversions if necessary), so I think `Model.to_bytes` needs to use something other than the current `ops.to_numpy`. Probably the cleanest API decision is add a byteorder kwarg to `to_numpy`, but the last time I looked I had trouble figuring out the numpy API to do this really cleanly for arbitrary formats...

**Comment by andrewsi-z** [comment]
>  Probably the cleanest API decision is add a byteorder kwarg to `to_numpy`, but the last time I looked I had trouble figuring out the numpy API to do this really cleanly for arbitrary formats...

If I'm interpreting this correctly I had the same observation looking at the API and thinking about to_numpy... I was thinking we'd have to inspect the host byteorder and adjust if needed based on that.  (We then started down the path initially of separating the support out to bigendian_ops and then I didn't need to implement anything anyway in to_numpy since numpy.save encodes byteorder for ndarray. And I wasn't so concerned about portability yet back to LE...)

**Comment by andrewsi-z** [comment]
> Glad to hear it! I think then the next question is how to handle the saved format, which I think we should probably always have as little endian. The problem is that `Ops.to_numpy` is also used within spacy in a few places for other reasons (numpy/cupy conversions if necessary), so I think `Model.to_bytes` needs to use something other than the current `ops.to_numpy`. Probably the cleanest API decision is add a byteorder kwarg to `to_numpy`, but the last time I looked I had trouble figuring out the numpy API to do this really cleanly for arbitrary formats...

Hi @adrianeboyd ... I just wanted to follow-up on this. Is this something it would be helpful for someone (me) to take a closer look at implementing?

**Comment by ptrcnull** [comment]
hi, is there any progress on this? i've stumbled over this issue trying to run tests on s390x and getting this failure:
```
=================================== FAILURES ===================================
___________________________ test_hash32_matches_mmh3 ___________________________
    def test_hash32_matches_mmh3():
        string = "hello world"
>       assert mmh3.hash(string) == murmurhash.mrmr.hash(string)
E       AssertionError: assert 1162266142 == 1586663183
E        +  where 1162266142 = <built-in function hash>('hello world')
E        +    where <built-in function hash> = mmh3.hash
E        +  and   1586663183 = <built-in function hash>('hello world')
E        +    where <built-in function hash> = <module 'murmurhash.mrmr' from 'mrmr.cpython-310-s390x-linux-gnu.so'>.hash
E        +      where <module 'murmurhash.mrmr' from 'mrmr.cpython-310-s390x-linux-gnu.so'> = murmurhash.mrmr
murmurhash/tests/test_against_mmh3.py:7: AssertionError
```

**Comment by adrianeboyd** [comment]
As far as we're aware this issue should be resolved as of murmurhash v1.0.7 / thinc v8.0.14 / spacy v3.3.0. If you run into any problems, please feel free to open a new issue.

**Comment by adrianeboyd** [comment]
You still need `thinc-bigendian-ops` to load models correctly, but I think you could currently use:

- murmurhash v1.0.7 (although there are no actual differences to v1.0.7.dev0)
- thinc v8.0.17
- spacy v3.3.1

You need thinc v8.0.14+ to get `thinc-apple-ops` to be detected automatically and spacy v3.3.0+ to get a fix for vector serialization.
